### ISSUE TITLE
parellel iteration in for

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-spec#1174]
Currently:
- To iterate the elements of two streams _in parellel_, you have to use `for ([x,y] in zipPairs(xs,ys))`.
- To iterate a stream with its indices, you have to use `for (i->x in xs.indexed)`.

The downside of this is the need to do an instantiation for each element.

I propose to let you write:

``` ceylon
for (x in xs, y in ys) { ... }
```

And

``` ceylon
for(i in xs.keys, x in xs) { ... }
```

Which will likely exhibit much superior performance.

**Comment by FroMage** [comment]
The problem with this is that most people will actually want parallel (as in asynchronous) execution of generators, like they use with `Future` composition using `for` comprehensions in Scala. I'm not sure it's possible to reconcile both synchronous parallel iteration and the composition behaviour of Scala's comprehensions in our case…

**Comment by gavinking** [comment]
I don't follow. What I'm proposing has nothing to do with Scala for comprehension. I didn't even notice until just now that the syntax looks similar.

**Comment by FroMage** [comment]
I know it has nothing to do with it. What I'm saying is that I think people will expect it to have to do with it ;)

**Comment by NiematojakTomasz** [comment]
> I didn't even notice until just now that the syntax looks similar.

And it means cartesian product. I thought so, some other language have such feature ; )

For other people who haven't googled it yet - http://docs.scala-lang.org/tutorials/tour/sequence-comprehensions.html

**Comment by gdejohn** [comment]
> And it means cartesian product.

I strongly agree with @gavinking that while there is slight potential for confusion with the Cartesian product upon seeing this syntax for the very first time, such a misunderstanding is only gonna happen at most once per person. This feature is simple and intuitive enough that a bit of documentation will suitably address the concerns here. This is the right way to do parallel iteration, and nested comprehensions are the right way to do the Cartesian product.

> I think people will expect it to have to do with [Scala for comprehensions].

I suppose Scala people would, but would anyone else? And how many Scala people are there? And anyway, I think this would be just one more difference from Scala that reflects *better* on Ceylon.

**Comment by FroMage** [comment]
To be clear, I like the current proposal personally, but I've heard feedback from Ceylon talks where people asked me explicitly if our comprehensions allowed for composition like Scala's and so I'm merely relaying user feedback. Perhaps there's a way to reconcile both? I doubt it because I think Scala's comprehensions are based on Gonads something, while ours are not.

**Comment by gavinking** [comment]
@FroMage we already have a syntax that works like Scala's comprehensions. That is, if I understand correctly, this:

```scala
for (x <- xs; y <-ys) yield new Point(x,y)
```

Is written like this in Ceylon:

```ceylon
for (x in xs) for (y in ys) Point(x,y)
```

Except that, as you mention, Scala's comprehensions are defined in terms of `flatMap()` whereas ours are defined in terms of iterators.

**Comment by gavinking** [comment]
Even after this discussion, and after reflection, I'm think I'm still in favor of doing this. But it's not for 1.2.

**Comment by luolong** [comment]
I'd cast my vote in favor of this proposal as well.
